src/allocation_engine/interval_tree.rs

```diff
@@ -359,6 +359,38 @@ impl InnerNode {
         self.update_cached_height();
         self.rotate()
     }
+
+    /// Insert a new node in the subtree. After the node is inserted the
+    /// tree will be balanced.
+    #[allow(dead_code)]
+    pub(crate) fn insert(
```

Comment:
How is this function going to be used?

There are 2 potential problems here:
1. Using recursion can generate a stack overflow, and it should be used with some defense mechanisms in place. One good option is to make sure you have a maximum possible depth. If the code can be written easily with iterations, than this should be the preferred method in production use cases.
2. Why is the `node_state` need to be passed for insert? Can we insert both allocated and free nodes?

Reply:
Indeed, we can also insert in tree a free node.

Reply:
I have addressed your comment regarding the recursion safety. The iterative code looks ugly and imho would be harder to maintain. I have computed how much a insert call should take on the stack and determined a max height. @raduiliescu had a really nice proposal to try to determine the stack size dynamically and to set the height on the run. While this is a nice proposal, I think we need an unsafe code block to do this.

Reply:
The `node_state` is still passed as a parameter. Can you remove it? Or otherwise explain why you need it?

Reply:
Updated the comment to explain why we need to have the `node_state` parameter